clang/lib/Serialization/ASTWriter.cpp

```diff
@@ -4628,6 +4628,12 @@ ASTFileSignature ASTWriter::WriteAST(Sema &SemaRef, StringRef OutputFile,
   WritingAST = true;
 
   ASTHasCompilerErrors = hasErrors;
+  bool trueHasErrors = SemaRef.PP.getDiagnostics().hasUncompilableErrorOccurred();
```

Comment:
So this logic seems wrong. You are on one hand asserting `ASTHasCompilerErrors == trueHasErrors` and yet checking if they are not equal.


Is `getDiagnostics().hasUncompilableErrorOccurred()` in `ASTUnit::serialize` not the same as `SemaRef.PP.getDiagnostics().hasUncompilableErrorOccurred()`?

Reply:
Hi @shafik, thanks for your time for the review.

The scenario here is that, at line 4630 you can see the `ASTHasCompilerErrors` is being set using `hasErrors` which is an input to the method here `ASTWriter::WriteAST()`.

And if you see the description of the issue reported, the user is purposely passing the `hasErrors` as `false` even when there is a compiler error in the input file, because of this clang is giving a crashing pch file.

So, if the `hasErrors` is not a valid value, I am setting it to the correct value. May be I can just remove all the unnecessary code and just directly assign `ASTHasCompilerErrors = SemaRef.PP.getDiagnostics().hasUncompilableErrorOccured();` which will be correct.

And assertion I have added to detect this in debug build early, may be I am wrong, sorry for the confusion, it should not behave differently in release and debug builds.

Let me know what you think of this?

Thanks
Rajkumar Ananthu.

Reply:
I would rather fix this by disallowing the user from passing `hasErrors` as an argument at all. The only callers of this API are all passing the same thing: `PP.getDiagnostics().hasUncompilableErrorOccured()`, so let's remove the parameter and use this directly. This makes the interface more simple and it fixes the crash; WDYT?

Reply:
Yeah @AaronBallman, that does sound better, I will remove the argument and update the pull request. Thanks for the review.

Reply:
Hi Team, Added a new commit by removing the argument `hasErrors` to the method `ASTWriter::WriteAST`.